src/defck.rs

```diff
@@ -0,0 +1,461 @@
+                            != self.grammar.provable_typecode()
+                        {
+                            // Non-provable typecodes are syntax axioms.
+                            // TODO Check that the axiom label does _not_ start with `df-`.
+                            let syntax_axiom = self.nset.lookup_label(stmt.label()).unwrap().atom;
+                            self.pending_syntax.push(syntax_axiom);
+                        } else if self.pending_syntax.is_empty() {
+                            // No definition to check, this is a regular axiom
+                            // TODO Check that the axiom label starts with `ax-`.
+                        } else if let Some(syntax) =
+                            self.ensure_frame_does_not_use_pending_syntax(&stmt, true)
+                        {
+                            // Definition, push it to the pending list for later processing
+                            self.pending_defn.push((stmt.address(), syntax));
+                        } else {
+                            // Also a regular axiom, it uses no new definitions
+                        }
+                    }
+                    StatementType::Provable => {
+                        if !self.pending_syntax.is_empty() {
+                            self.flush_pending_definitions();
+                            self.ensure_frame_does_not_use_pending_syntax(&stmt, false);
+                        }
+                    }
+                    _ => {}
+                }
+            }
+        }
+
+        self.flush_pending_syntax();
+        for (_, addr, delayed_diag) in self.pending_primitive.drain(..) {
+            self.result.diagnostics.push((addr, delayed_diag));
+        }
+    }
+
+    fn ensure_frame_does_not_use_pending_syntax(
+        &mut self,
+        stmt: &StatementRef<'_>,
+        except_for_one: bool,
+    ) -> Option<Label> {
+        let res = self.ensure_statement_does_not_use_pending_syntax(stmt, except_for_one);
+        if let Some(frame) = self.scope.get(stmt.label()) {
+            for hyp in &*frame.hypotheses {
+                if let Hyp::Essential(addr, _) = *hyp {
+                    let stmt = self.db.statement_by_address(addr);
+                    self.ensure_statement_does_not_use_pending_syntax(&stmt, false);
+                }
+            }
+        }
+        res
+    }
+
+    fn ensure_statement_does_not_use_pending_syntax(
+        &mut self,
+        stmt: &StatementRef<'_>,
+        except_for_one: bool,
+    ) -> Option<Label> {
+        let mut the_def = None;
+        if let Some(f) = self.stmts.get_formula(stmt) {
+            for (label, _) in f.labels_iter().filter(|(_, var)| !var) {
+                if let Some(i) = self.pending_syntax.iter().position(|&x| x == label) {
+                    self.pending_syntax.swap_remove(i);
+                    if except_for_one && the_def.map_or(true, |l| l == label) {
```

Comment:
If I understand well, this means that the one pending syntax allowed,`the_def`, might appear twice or more in the definition.
This would allow for e.g. recursive definitions.
Is that wanted? I thought definitions could only be in the form `A = B`, where `A` does not appear in `B`.

Reply:
This is just the first pass, this does not implement the actual definition check. For definitions via justification though you do want to allow the LHS to appear multiple times in the formula.

Reply:
Ok, so it will be in the actual definition check that we'll check that the definiendum appears only once.